posthog/clickhouse/client/tracing.py

```diff
@@ -0,0 +1,116 @@
+import logging
+from functools import wraps
+from time import perf_counter
+
+from opentelemetry import trace
+from opentelemetry.trace import Status, StatusCode
+
+from posthog.clickhouse.client.connection import Workload, ClickHouseUser
+from posthog.settings import CLICKHOUSE_DATABASE, CLICKHOUSE_HOST
+
+logger = logging.getLogger(__name__)
+
+
+def trace_clickhouse_query_decorator(func):
+    """
+    Decorator to add ClickHouse query tracing to sync_execute function.
+    This decorator handles the complex tracing requirements including:
+    - Workload changes during retries
+    - Result tracking
+    - Exception handling
+    - Execution time measurement
+    """
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        # Extract parameters for tracing - handle both positional and keyword args
+        query = args[0] if args else kwargs.get("query")
+        args_param = kwargs.get("args")
+        initial_workload = kwargs.get("workload", Workload.DEFAULT)
+        team_id = kwargs.get("team_id")
+        readonly = kwargs.get("readonly", False)
+        ch_user = kwargs.get("ch_user", ClickHouseUser.DEFAULT)
+
+        # Handle positional arguments for sync_execute signature
+        # sync_execute has: query, args=None, settings=None, with_column_types=False, flush=True, *, workload, team_id, readonly, sync_client, ch_user
+        if len(args) > 1:
+            args_param = args[1]
+
+        # The keyword-only arguments (after *) must be passed as kwargs
+        # So we only extract from kwargs for these parameters
+
+        tracer = trace.get_tracer(__name__)
+
+        with tracer.start_as_current_span("clickhouse.query") as span:
+            if span.is_recording():
```

Comment:
aren't all tracer functions noops anyway if !is_recording?

if not, shouldn't the is_recording check be at the start of the wrapper? (`if not span.is_recording(): return func(*args, **kwargs)`)